### ISSUE TITLE
Adding Documentation Notes

### ISSUE THREAD

**Issue by trisongz** [ThilinaRajapakse/simpletransformers#295]
Hi - I recently started using simpletransformers to prototype some models due to the simplicity of the library and it's been a great experience so far.

I wanted to share some of the issues I ran into and how I got around them.

**Language Modeling** 

When training a model and tokenizer from scratch, it's not clear that you should have both the tokenizer configs as well as the model configs in the same args when it's passed through when calling 

`model = ClassificationModel('gpt2' None, args=args)`

and then calling

`model.train_tokenizer(train_files='/folder/trainfile.txt', output_dir='/folder/tokenizeroutput/')`

prior to calling 

`model.train_model(train_file='/folder/trainfile.txt',  eval_file='/folder/evalfile.txt')`

For GPT2, it should be noted that mlm for the config should be False as well, and having a base tokenizer class for tokenizer_name (since model = None doesn't pass a model config through).

My configuration prior to tokenization of a new from scratch model (GPT2)

```
train_args={
    'reprocess_input_data': True,
    'overwrite_output_dir': True,
    'evaluate_during_training': True,
    'evaluate_during_training_steps': 10000,
    'dataset_type': 'simple',
    'mlm': False,
    'num_train_epochs': 5,
    'logging_steps': 200,
    'best_model_dir': '/content/model/best/',
    'output_dir': '/content/model/',
    'tokenizer_name': 'gpt2',
    'train_files': '/content/train.txt',
    'learning_rate': 1e-5,
    'max_seq_length': 60,
    'vocab_size': 52000,
}

model = LanguageModelingModel('gpt2', None, args=train_args)
model.train_tokenizer(train_files='/content/train.txt', output_dir='/content/tokenizer/')
model.train_model(train_file='/content/train.txt', eval_file='/content/eval.txt')
```

Afterwards, when calling the tokenizer/pre-trained model again (if session restarts)

```
train_args={
    'reprocess_input_data': True,
    'overwrite_output_dir': True,
    'evaluate_during_training': True,
    'evaluate_during_training_steps': 10000,
    'dataset_type': 'simple',
    'mlm': False,
    'num_train_epochs': 5,
    'logging_steps': 200,
    'best_model_dir': '/content/model/best/',
    'output_dir': '/content/model/',
    'tokenizer_name': '/content/tokenizer/',
    'train_files': '/content/train.txt',
    'learning_rate': 1e-5,
    'max_seq_length': 60,
    'vocab_size': 52000,
}

model = LanguageModelingModel('gpt2', '/content/model/best/', args=train_args)
model.train_model(train_file='/content/train.txt', eval_file='/content/eval.txt')
```

**ClassificationModel - Regression**

One of the errors I had kept running into was when using pandas to read from a CSV, if the text column contained an int or a float within the text, it would throw an error as the tokenizer couldn't parse it properly. The quick fix for this is declaring the column type using dtype when loading in the csv file.

df = pd.read_csv("/content/data.csv", sep='|', dtype={'text_a': str, 'text_b': str, 'labels': int})
df = pd.DataFrame(df, columns=['text_a', 'text_b, 'labels'])

I still run into issues when using float for labels however, so I defaulted to using integers.

**Quality of Life Stuff**

When using Pandas to load a single csv, a useful function for splitting the dataset without having to have different datasets is by using sklearn's train_test_split - found from https://github.com/lvwerra/trl/blob/master/nbs/03-bert-imdb-training.ipynb

```
import sklearn
from sklearn.model_selection import train_test_split

df = pd.read_csv("/content/data.csv", sep='|', dtype={'text_a': str, 'text_b': str, 'labels': int})
df = pd.DataFrame(df, columns=['text_a', 'text_b, 'labels'])
train_df, eval_df = train_test_split(df, test_size=0.2)
train_df.head()
```

If you use wandb for logging training, it can sometimes be really verbose, and stall/crash your notebook. I had initially tried to set the env to WANDB_SILENT=True as per their docs, but it didn't work. Easiest way to fix this is with

```
import logging
logging.basicConfig(level=logging.INFO)
wandb_logger = logging.getLogger("wandb")
wandb_logger.setLevel(logging.WARNING)
```

Question - when working with larger datasets (2gb), I had kept running into an issue where during data preprocessing, it had kept crashing the notebook due to out of ram.

- Task: Sequence Classification
- Dataset: csv, loaded with pandas
- Model: bert-base-cased
- Batch Size: 4/8
- VM: Colab-GPU - 25.5 GB RAM with P100

I had tried to reprocess_input_data = False but it would still use all the RAM before crashing, so I ended up using a smaller dataset. Any tips for getting around this?

**Comment by ThilinaRajapakse** [comment]
The documentation for the Language Model training definitely leaves a lot to be desired. It, and perhaps the code itself, is still a bit of a work in progress. Thank you for pointing out what needs to be fixed as it helps a lot!

That is odd. The tokenizer should definitely be able to parse numbers within the text, but I wouldn't be surprised if things went haywire if the text is _only_ a number.

Yes, you can certainly use sklearn to split a pandas df if needed. If you are mentioning this because of some minimal examples where I just use the training data itself for testing, that's because I really am lazy. Too lazy to write those 2 extra lines at times. :sweat_smile: 

The data preprocessing is quite memory inefficient. That portion of the code was taken directly from the transformers library for the most part, and I simply added multiprocessing on top to speed up the process. I did manage to cut down the memory usage for the language modelling tasks where I wrote most of the processing code myself. It's possible to port that over for the other tasks so I'll see what I can do. If you are using Colab, I believe you can request a VM with more memory if needed.

**Comment by trisongz** [comment]
I definitely understand - maintaining documentation and improving the code base are two equally difficult and time consuming tasks. Also would note a new bug for LM - if max_steps aren't defined and you're using wandb - it errors out the training during evaluation as it requires a 'finite' range.

I currently use Colab Pro and use the highest mem GPU-VM available (25.5GB), which is where I ran into the problem. 

I've been working a bit with the new [PyTorch lightning](https://github.com/PytorchLightning/pytorch-lightning) project and it seems like it may be beneficial, as one of the great things it does is handle a lot of the things data preprocessing as part of the batching and forward step, rather than requiring all upfront pre-processing and holding it in memory. I was planning to start porting some of the functions/models from simpletransformers over to lightning once I had a bit of time.
